File: benchmark/sieve.py
#!/usr/bin/env python3
# Sieve of Eratosthenes benchmark for Python
# Counts primes up to LIMIT using the sieve algorithm
# Reference: π(10,000,000) = 664,579
#
# How to run:
#   make benchmark-sieve
#
# Or manually:
#   python3 benchmark/sieve.py

import time


def sieve_count(limit: int) -> int:
    # Use bytearray for better performance than list of bools
    is_prime = bytearray(limit + 1)
    for i in range(limit + 1):
        is_prime[i] = 1

    # 0 and 1 are not prime
    is_prime[0] = 0
    is_prime[1] = 0

    # Sieve: mark multiples of each prime as not prime
    p = 2
    while p * p <= limit:
        if is_prime[p]:
            # Mark all multiples of p starting from p*p
            for multiple in range(p * p, limit + 1, p):
                is_prime[multiple] = 0
        p += 1

    # Count primes
    count = sum(is_prime)

    return count


def main():
    limit = 10000000

    start = time.perf_counter_ns()
    count = sieve_count(limit)
    end = time.perf_counter_ns()

    elapsed_ns = end - start
    elapsed_ms = elapsed_ns // 1000000

    print(f"Sieve prime count up to {limit}: {count}")
    print(f"Elapsed: {elapsed_ms} ms")


if __name__ == "__main__":
    main()

File: benchmark/sieve/sieve.rb
#!/usr/bin/env ruby
# Sieve of Eratosthenes benchmark for Ruby
# Counts primes up to LIMIT using the sieve algorithm
# Reference: π(10,000,000) = 664,579
#
# How to run:
#   mise run benchmark-sieve
#
# Or manually:
#   ruby benchmark/sieve.rb

def sieve_count(limit)
  # Use array of booleans
  is_prime = Array.new(limit + 1, true)

  # 0 and 1 are not prime
  is_prime[0] = false
  is_prime[1] = false

  # Sieve: mark multiples of each prime as not prime
  p = 2
  while p * p <= limit
    if is_prime[p]
      # Mark all multiples of p starting from p*p
      multiple = p * p
      while multiple <= limit
        is_prime[multiple] = false
        multiple += p
      end
    end
    p += 1
  end

  # Count primes
  is_prime.count(true)
end

def main
  limit = 10_000_000

  start = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  count = sieve_count(limit)
  elapsed_ns = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - start
  elapsed_ms = elapsed_ns / 1_000_000

  puts "Sieve prime count up to #{limit}: #{count}"
  puts "Elapsed: #{elapsed_ms} ms"
end

main
